tools/udp_rcvbuf_errors.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+#
+# udp_rcvbuf_errors.py	UDP RcvbufErrors analysis tool.
+#
+# Prints out information for UDP packets which were dropped
+# a result of the socket receive buffer being full.
+#
+# USAGE: udp_rcvbuf_errors
+
+# Copyright (c) 2019 Cloudflare, Inc.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 15-Jan-2019 Thomas Lefebvre Created this.
+
+from bcc import BPF
+import ctypes as ct
+import argparse
+from time import strftime
+from socket import inet_ntop, AF_INET, AF_INET6
+from struct import pack
+
+examples = """examples:
+    ./udp_rcvbuf_errors           # trace UDP RcvbufErrors kernel drops
+"""
+parser = argparse.ArgumentParser(
+    description="Trace UDP RcvbufErrors drops by the kernel",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+args = parser.parse_args()
+debug = 0
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/udp.h>
+#include <uapi/linux/ip.h>
+#include <net/sock.h>
+
+struct ipv4_data_t {
+    u32 saddr;
+    u32 daddr;
+    u16 sport;
+    u16 dport;
+    u32 rmem_alloc;
+    u32 rcvbuf;
+};
+BPF_PERF_OUTPUT(ipv4_events);
+
+struct ipv6_data_t {
+    unsigned __int128 saddr;
+    unsigned __int128 daddr;
+    u16 sport;
+    u16 dport;
+    u32 rmem_alloc;
+    u32 rcvbuf;
+};
+BPF_PERF_OUTPUT(ipv6_events);
+
+static struct udphdr *skb_to_udphdr(const struct sk_buff *skb)
+{
+    // unstable API. verify logic in udp_hdr() -> skb_transport_header().
+    return (struct udphdr *)(skb->head + skb->transport_header);
+}
+
+static inline struct iphdr *skb_to_iphdr(const struct sk_buff *skb)
+{
+    // unstable API. verify logic in ip_hdr() -> skb_network_header().
+    return (struct iphdr *)(skb->head + skb->network_header);
+}
+
+int kprobe__udp_queue_rcv_skb(struct pt_regs *ctx, struct sock *sk, struct sk_buff *skb) {
+    if (sk->sk_rmem_alloc.counter <= sk->sk_rcvbuf)
+    {
```

Comment:
put `{` in previous linee?
Also, could you give a little more explanation how do you come up with the above condition? Any caveat here (e.g., any possible missed counting?) Some comments will help people understand the approach applied here? Also it seems `udp_queue_rcv_skb` is a static function and it could get inlined. It it does get inlined, the counting will not be accurate. Maybe some warning in comments as well.